shared/src/ledger/ibc/connection.rs

```diff
@@ -0,0 +1,423 @@
+            // returns None even if DB read fails
+            _ => None,
+        }
+    }
+
+    fn connection_counter_pre(&self) -> u64 {
+        let path = COUNTER_PATH.to_owned();
+        let key = Key::ibc_key(path)
+            .expect("Creating a key for a connection counter failed");
+        match self.ctx.read_pre(&key) {
+            Ok(Some(value)) => storage::types::decode(&value)
+                .expect("converting a connection counter shouldn't failed"),
+            _ => {
+                tracing::error!("connection counter should exist");
+                unreachable!();
+            }
+        }
+    }
+}
+
+impl<'a, DB, H> ConnectionReader for Ibc<'a, DB, H>
+where
+    DB: 'static + storage::DB + for<'iter> storage::DBIter<'iter>,
+    H: 'static + StorageHasher,
+{
+    fn connection_end(&self, conn_id: &ConnectionId) -> Option<ConnectionEnd> {
+        let path = Path::Connections(conn_id.clone()).to_string();
+        let key = Key::ibc_key(path)
+            .expect("Creating a key for a connection end failed");
+        match self.ctx.read_post(&key) {
+            Ok(Some(value)) => ConnectionEnd::decode_vec(&value).ok(),
+            // returns None even if DB read fails
+            _ => None,
+        }
+    }
+
+    fn client_state(&self, client_id: &ClientId) -> Option<AnyClientState> {
+        ClientReader::client_state(self, client_id)
+    }
+
+    fn host_current_height(&self) -> Height {
+        let epoch = self.ctx.storage.get_block_epoch().0.0;
+        let height = self.ctx.storage.get_block_height().0.0;
+        Height::new(epoch, height)
+    }
+
+    fn host_oldest_height(&self) -> Height {
+        let epoch = Epoch::default().0;
+        let height = BlockHeight::default().0;
+        Height::new(epoch, height)
+    }
+
+    fn commitment_prefix(&self) -> CommitmentPrefix {
+        let addr = Address::Internal(InternalAddress::Ibc);
+        let bytes = addr
+            .raw()
+            .try_to_vec()
+            .expect("Encoding an address string shouldn't fail");
+        CommitmentPrefix::from(bytes)
+    }
+
+    fn client_consensus_state(
+        &self,
+        client_id: &ClientId,
+        height: Height,
+    ) -> Option<AnyConsensusState> {
+        self.consensus_state(client_id, height)
+    }
+
+    fn host_consensus_state(
+        &self,
+        _height: Height,
+    ) -> Option<AnyConsensusState> {
+        self.ctx
+            .storage
+            .get_block_header()
+            .0
+            .map(|h| TendermintConsensusState::from(h).wrap_any())
+    }
+
+    fn connection_counter(&self) -> u64 {
+        let path = COUNTER_PATH.to_owned();
+        let key = Key::ibc_key(path)
+            .expect("Creating a key for a connection counter failed");
+        match self.ctx.read_post(&key) {
+            Ok(Some(value)) => storage::types::decode(&value)
+                .expect("converting a connection counter shouldn't failed"),
```

Comment:
I think we have to handle the decoding error without `expect` to avoid panic on malicious state update

Reply:
Thanks! Fixed.
I'll fix that of the client counter in another PR.